EIPS/eip-7990.md

```diff
@@ -0,0 +1,302 @@
+
+## Motivation
+
+Current Ethereum contracts requiring dynamic code execution must use `DELEGATECALL` to external contracts containing the desired bytecode. This approach has several limitations:
+
+1. **Cold address access overhead**: `DELEGATECALL` incurs a 2600 gas cold address access penalty when the target contract hasn't been accessed in the current transaction, plus the 100 gas base call cost
+2. **Address dependency**: Requires managing and accessing external contract addresses, creating additional state dependencies
+3. **Deployment requirements**: Dynamic code must be pre-deployed as contracts, consuming storage and deployment gas (~32,000 gas + ~200 gas per byte)
+4. **Address space pollution**: Each piece of dynamic code requires a unique contract address, leading to blockchain state bloat
+5. **Architectural complexity**: Managing multiple deployed contracts for dynamic execution adds significant complexity to contract systems
+
+Common patterns that would benefit include:
+
+- Factory contracts calling multiple dynamic computation libraries
+- Upgradeable proxy contracts with frequently changing implementations
+- Meta-programming patterns where code is generated and executed at runtime
+- Mathematical computation libraries that generate optimized bytecode
+- One-time execution of complex calculations without permanent on-chain storage
+
+The `RUNCODE` opcode addresses these limitations by allowing direct execution of bytecode from memory within the current execution context, eliminating address access costs and deployment requirements while providing the same functionality as `DELEGATECALL`.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119) and [RFC 8174](https://www.rfc-editor.org/rfc/rfc8174).
+
+### RUNCODE Opcode
+
+- **Opcode**: `0xF6`
+- **Name**: `RUNCODE`
+- **Gas Cost**: 100 base gas (equivalent to DELEGATECALL base cost without address access)
+- **Stack Input**: `[gas, codeOffset, codeLength, argsOffset, argsLength, retOffset, retLength]`
+- **Stack Output**: `[success]`
+
+### Behavior
+
+1. **Input Parameters**: Takes seven stack parameters (identical to DELEGATECALL interface):
+   - `gas`: Amount of gas to provide to the dynamic execution
+   - `codeOffset`: Memory offset where bytecode starts
+   - `codeLength`: Length of bytecode in bytes
+   - `argsOffset`: Memory offset of input data to pass to dynamic code
+   - `argsLength`: Length of input data in bytes
+   - `retOffset`: Memory offset to store return data
+   - `retLength`: Maximum length of return data to copy back
+
+2. **Execution Context**:
+
```

Comment:
do consistent formatting, for e.g you have a new line inserted  here but not at other points